pkg/controller/argocdmetrics/argocdmetrics_controller.go

```diff
@@ -0,0 +1,351 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	readRoleNameFormat        = "%s-read"
+	readRoleBindingNameFormat = "%s-prometheus-k8s-read-binding"
+	alertRuleName             = "gitops-operator-argocd-alerts"
+)
+
+var logs = logf.Log.WithName("controller_argocd_metrics")
+
+type ArgoCDMetricsReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// blank assignment to verify that ReconcileArgoCDRoute implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ArgoCDMetricsReconciler{}
+
+// Add creates a new ArgoCD Metrics Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and start it when the
+// Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func add(mgr manager.Manager, reconciler reconcile.Reconciler) error {
+	reqLogger := logs.WithValues()
+
+	reqLogger.Info("Creating ArgoCD metrics controller")
+	c, err := controller.New("argocd-metrics-controller", mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+
+	reqLogger.Info("Watching for ArgoCD instance creation")
+	err = c.Watch(&source.Kind{Type: &argoapp.ArgoCD{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ArgoCDMetricsReconciler{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+func (r ArgoCDMetricsReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := logs.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ArgoCD Metrics")
+
+	namespace := corev1.Namespace{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: request.Namespace}, &namespace)
+	if err != nil {
+		reqLogger.Error(err, "Error getting namespace",
+			"Namespace", request.Namespace)
+		return reconcile.Result{}, err
+	}
+	_, exists := namespace.Labels["openshift.io/cluster-monitoring"]
+	if !exists {
+		if namespace.Labels == nil {
+			namespace.Labels = make(map[string]string)
+		}
+		namespace.Labels["openshift.io/cluster-monitoring"] = "true"
+		err = r.client.Update(context.TODO(), &namespace)
+		if err != nil {
+			reqLogger.Error(err, "Error updating namespace",
+				"Namespace", namespace.Name)
+			return reconcile.Result{}, err
+		}
+	} else {
+		reqLogger.Info("Namespace already has cluster-monitoring label",
+			"Namespace", namespace.Name)
+	}
+
+	// Create role to grant read permission to the openshift metrics stack
+	err = r.createReadRoleIfAbsent(request.Namespace, reqLogger)
```

Comment:
Could have used the `view` cluster role but that probably isn't a good idea since `view` cluster role grants a lot more.

Reply:
Also set `ownerReferences` to the corresponding `ArgoCD` CR ?